Ai2thor wrapper

beduffy opened the pull request:
Added interface with 12 actions and simple mug picking task
Added random walk.

beduffy committed:
Added barebones wrapper for ai2thor. As simple as possible but clean

Task is to pick up cups, 1 reward per cup. Allowed open and close objects.
Much cleaner action space. Still one silent bug to fix.

beduffy committed:
Added Random walk

beduffy committed:
Fixed silent crash bug on opening/closing objects and many more changes

Added object instance check for bounding box for pickup
Fixed episode length check
Reading in config file for pickupable objects and looping through those
Moved env stuff into own folder with config.ini as sibling and read with
config parser
Added timing and step+episode number printing to random walk

beduffy committed:
Added max episode length to random walk

Added acceptable receptacles
Removed one raise_for_failure
Renamed vars to not be mug specific.
Reading config file with strips for spaces.

fernandotorch commented on ai2thor_wrapper/config.ini:
```diff
@@ -0,0 +1,5 @@
+[ENV_SPECIFIC]
+;PICKUPABLE_OBJECTS=Mug
+PICKUPABLE_OBJECTS=Mug,Apple,Book,Bowl,Plate,Tomato,Pan,Fork,Knife,Spoon,Potato,Lettuce,Egg,ButterKnife
+ACCEPTABLE_RECEPTACLES=CounterTop,TableTop,Sink
```
I would rename those two to these:
PICKUP_OBJECTS. It sounds more natural to just say pick up.
RECEPTACLE_OBJECTS. Also a property of a specific group of objects, we can use objects in the name as convention

beduffy replied:
This is the only point in my life where I can rejoice in finally being able to call something ACCEPTABLE_RECEPTACLES and you want it renamed?

PICKUP_OBJECTS sure

fernandotorch commented on ai2thor_wrapper/config.ini:
```diff
@@ -0,0 +1,5 @@
+[ENV_SPECIFIC]
```
Change config.ini to config.ini.example or something like that, then never commit your config to the main repo. Add to gitignore, this one should be only a clean guide for people starting on the repo

beduffy replied:
done

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,194 @@
+import skimage.color, skimage.transform
+import ai2thor.controller
+
+ALL_POSSIBLE_ACTIONS = [
```
Rename to ALLOWED_ACTIONS or POSSIBLE_ACTIONS.

fernandotorch commented on examples/random_walk.py:
```diff
@@ -0,0 +1,22 @@
+    env = ThorWrapperEnv(max_episode_length=max_episode_length)
+    for episode in range(20):
+        start = time.time()
+        s = env.reset()
```
initial_state, state_ini or sth similar would be preferred to "s". In general I would avoid single letter variables

fernandotorch commented on examples/random_walk.py:
```diff
@@ -0,0 +1,22 @@
+        start = time.time()
+        s = env.reset()
+        for t in range(max_episode_length):
+            a = random.randint(0, len(env.ACTION_SPACE) - 1)
```
use random.choice(env.ACTION_SPACE) and call a "action"

beduffy replied:
good call in general but i don't have a list of actions to do choice from.
`random.choice(range(env.NUM_ACTIONS))` might work though

fernandotorch commented on examples/random_walk.py:
```diff
@@ -0,0 +1,22 @@
+        s = env.reset()
+        for t in range(max_episode_length):
+            a = random.randint(0, len(env.ACTION_SPACE) - 1)
+            s, r, done = env.step(a)
```
state, reward

fernandotorch commented on examples/random_walk.py:
```diff
@@ -0,0 +1,22 @@
+    for episode in range(20):
+        start = time.time()
+        s = env.reset()
+        for t in range(max_episode_length):
```
range(1, max_episode_length) will give the expected behaviour right away, and lets you remove all the t+1 below

beduffy replied:
Fernando, the reason i have t+1 is because it looks weird to see 0-99

AND

if we did `range(1, max_episode_length)` that would be 1 less. It has to be:
`range(1, max_episode_length + 1)`
And now that's a bit confusing and ugly

Soooo. Stick with t+1?

beduffy committed:
Added README.md and requirements.txt, moved random_walk.py to root

Renamed config.ini to config_example.ini to stress ability to change
this file.
Added which object was placed in print.
A few var renames

beduffy committed:
Added unittest to checks to see if the wrapper works and nothing breaks.

Useful for continuous deployment and keeping master stable. Also, we
check how much time 100 steps takes in the environment takes. Final assert
checks if max_episode_length is equal to the number of steps taken and no off-by-one errors.

fernandotorch commented on README.md:
```diff
@@ -1 +1,23 @@
+# ai2thor-experiments
+
+We created an OpenAI gym style wrapper for [ai2thor](http://ai2thor.allenai.org/) and show a random walk example below.
+We intend to experiment with different model-free and model-based algorithms within this repo with natural language 
```
using natural language... It is being picky, but you used "wit" several times
```suggestion
We intend to experiment with different model-free and model-based algorithms within this repo using natural language 
```

fernandotorch commented on ai2thor_wrapper/config_example.ini:
```diff
@@ -0,0 +1,5 @@
+[ENV_SPECIFIC]
+;PICKUPABLE_OBJECTS=Mug
```
Modify the commented line as well because it is confusing otherwise
```suggestion
;PICKUP_OBJECTS=Mug
```

fernandotorch commented on README.md:
```diff
@@ -1 +1,23 @@
+
+# How to run
+
+`python random_walk.py`
```
Go to `examples` folder and execute `random_walk.py`

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+        config_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), config_path)
+        self.config = configparser.ConfigParser()
+        config_output = self.config.read(config_path)
+        if len(config_output) == 0:
```
```suggestion
        if len(config_output):
```

fernandotorch replied:
or even better
```suggestion
        if config_output:
```

beduffy replied:
dangerous. Are you not aware of Falsy and falsy dangerous stuff in multiple languages (or maybe i've just been tainted by JavaScript).
But Python has a few gotchas about this

fernandotorch replied:
According to PEP8:
```
For sequences, (strings, lists, tuples), use the fact that empty sequences are false.

Yes: if not seq:
     if seq:

No:  if len(seq):
     if not len(seq):
```

Source: https://www.python.org/dev/peps/pep-0008/

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+        self.config = configparser.ConfigParser()
+        config_output = self.config.read(config_path)
+        if len(config_output) == 0:
+            print('No config file found at: {}. Exiting'.format(config_path))
```
No print and exit. Raise it!
```suggestion
           raise ValueError('No config file found at: {}. Exiting'.format(config_path))
```

beduffy replied:
yep, i knew you'd say that haha

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+        config_output = self.config.read(config_path)
+        if len(config_output) == 0:
+            print('No config file found at: {}. Exiting'.format(config_path))
+            sys.exit()
```
```suggestion
```

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+
+        # action dictionary from int to dict action to pass to event.controller.step()
+        self.ACTION_SPACE = {}
+        for idx, action_str in enumerate(POSSIBLE_ACTIONS):
```
This data structure feels strange to me
```suggestion
        self.ACTION_SPACE = [{idx: action_str} for idx, action_str in enumerate(POSSIBLE_ACTIONS)]
```

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+        for idx, action_str in enumerate(POSSIBLE_ACTIONS):
+            self.ACTION_SPACE[idx] = dict(action=action_str)
+
+        if not interaction:
```
I would solve the interaction thing differently. Create two lists: moving_actions and interaction_actions. Then:
```suggestion
        possible_actions = moving_actions.copy()
        if interaction:
            possible_actions.extend(interaction_actions)
```
And continue with the dict comprehension as before

beduffy replied:
seems a bit more complex rather than just "removing interaction actions if no interaction allowed" no?

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+                if self.ACTION_SPACE[idx]['action'] in interaction_actions:
+                    self.ACTION_SPACE.pop(idx)
+
+        self.NUM_ACTIONS = len(self.ACTION_SPACE.keys())
```
```suggestion
        self.NUM_ACTIONS = len(self.ACTION_SPACE)
```

beduffy replied:
dangerous and not explicit maybe?

fernandotorch replied:
What is not explicit about it? It is explained in python docu 
```
len(s)
Return the length (the number of items) of an object. The argument may be a sequence (such as a string, bytes, tuple, list, or range) or a collection (such as a dictionary, set, or frozen set).
```
Source: https://docs.python.org/3/library/functions.html#len

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+        self.action_space = self.NUM_ACTIONS
+
+        # acceptable objects taken from config.ini file. Stripping to allow spaces
+        self.pickup_object_types = [x.strip() for x in self.config['ENV_SPECIFIC']['PICKUP_OBJECTS'].split(',')]
```
If interaction somewhere here? or put this up also when you select the interactions. If I say no interactions and then you list me pickable objects I am going to be confused

beduffy replied:
yep good idea

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+
+    def step(self, action_int):
+        if self.ACTION_SPACE[action_int]['action'] == 'PickupObject':
+            if len(self.event.metadata['inventoryObjects']) == 0:
```
```suggestion
            if self.event.metadata['inventoryObjects']:
```

beduffy replied:
I really don't think this is pythonic and also can be quite dangerous.

fernandotorch replied:
Why don't you think so? It is explicit in python docu. This is the pythonic way

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+    def step(self, action_int):
+        if self.ACTION_SPACE[action_int]['action'] == 'PickupObject':
+            if len(self.event.metadata['inventoryObjects']) == 0:
+                for o in self.event.metadata['objects']:
```
Give me at least 3 letters. I hate having one, I forget what the name means
```suggestion
                for obj in self.event.metadata['objects']:
```

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+            if len(self.event.metadata['inventoryObjects']) == 0:
+                for o in self.event.metadata['objects']:
+                    # loop through objects that are visible, pickupable and there is a bounding box visible
+                    if o['visible'] and o['pickupable'] and o['objectType'] in self.pickup_object_types \
```
```suggestion
                    if o['visible'] and 
                       o['pickupable'] and 
                       o['objectType'] in self.pickup_object_types and 
                       o['objectId'] in self.event.instance_detections2D:
```

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+                        if object_type == self.current_object_type:
+                            self.goal_objects_collected_and_placed.append(object_id)
+                        print('Picked up', self.event.metadata['inventoryObjects'])
+                        break
```
why break?

beduffy replied:
if you have pickedup or put down something, you have succeeded. No need to loop through the rest of the objects

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+                        print('Picked up', self.event.metadata['inventoryObjects'])
+                        break
+        elif self.ACTION_SPACE[action_int]['action'] == 'PutObject':
+            if len(self.event.metadata['inventoryObjects']) > 0:
```
```suggestion
            if self.event.metadata['inventoryObjects']:
```

beduffy replied:
I really don't think this is pythonic and also can be quite dangerous.

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+                        break
+        elif self.ACTION_SPACE[action_int]['action'] == 'PutObject':
+            if len(self.event.metadata['inventoryObjects']) > 0:
+                for o in self.event.metadata['objects']:
```
pleaaaase
```suggestion
                for obj in self.event.metadata['objects']:
```

beduffy replied:
sure. 3 letters works if it's clearer
but 
"a" to "act" might be actual less clear

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+            if len(self.event.metadata['inventoryObjects']) > 0:
+                for o in self.event.metadata['objects']:
+                    # loop through receptacles
+                    if o['visible'] and o['receptacle'] and \
```
```suggestion
                    if o['visible'] and 
                    o['receptacle'] and
                    o['objectType'] in self.receptacle_object_types and
                    len(o['receptacleObjectIds']) < o['receptacleCount']:
```

beduffy replied:
that really doesn't look like PEP8 but maybe it is haha. I only did it here though.

Notice there is only 1 horizontal space below that line  and we've broken magic 4 tabs

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+        elif self.ACTION_SPACE[action_int]['action'] == 'OpenObject':
+            for o in self.event.metadata['objects']:
+                # loop through objects that are visible, openable, closed
+                if o['visible'] and o['openable'] and not o['isopen'] and o['objectType'] == 'Microwave' \
```
ouch, hard coded Microwave? If so, why loop over all event.metadata['objects'] in the first place?

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+        self.t += 1
+        state = self.preprocess(self.event.frame)
+        reward = self.calculate_reward(self.done)
+        self.done = self.is_episode_finished()
```
Please do that in the function name as well
```suggestion
        self.done = self.episode_finished()
```

beduffy replied:
yeah, i hated that function name

fernandotorch commented on ai2thor_wrapper/envs.py:
```diff
@@ -0,0 +1,195 @@
+        return state, reward, self.done
+
+    def preprocess(self, img):
+        img = skimage.transform.resize(img, self.resolution)
```
you should wrap the resizing in a function as you did with rgb2gray

beduffy replied:
nah that would be overkill, it's just 1 line? rgb2gray is an actual numpy statement and calculation which is different

beduffy committed:
Added most of Fernando's changes

typos, better var names (e.g. current task object, obj, episode_finished)
Put random_walk back into examples but sys.path.append
Updated README.md with requirements and added 2 __init__.pys
dont print interaction objects if false
Added openable objects

beduffy committed:
Added 3 ai2thor examples as tests and modified 3rd to be a stress test and

threading and fps/speed test (for different levels of extra info: class, object and depth)

beduffy committed:
Renamed ACTION_SPACE to ACTION_DICT and made separate list for space

this allows random.choice in random_walk. Trying to isolate put down bug.

fernandotorch committed:
Fix code style and prints

Using pylint for code style.
The multithreaded example had different steps printed as the comments explained.

fernandotorch committed:
Add openAI gym interfaces

Redefined the environment according to gym docu
Clean and general refactor without changes
Code is not usable at this point, will refactor the functions and even move some outside

fernandotorch committed:
Split Env and Task classes

Added a task factory class in an indpendent script for modularity.
Simplified parts of env class, but not completely.
Still non working code

fernandotorch committed:
Fix env wrapper

Now environment and task classes are working for the dummy config. The task rewards are just a dummy function and not well implemented yet.

fernandotorch committed:
Improve readability of action message

Changed the dictionary with all info from the object to just its name so it looks better

fernandotorch committed:
Separate Task factory from BaseTask

Now both are independent classes. Tasks inherit from BaseTask but are selected with the factory

fernandotorch committed:
Refactor Env

Changed constructor input, internal variable names and added inline comments for the action handling. Separated image processing to its own script.

fernandotorch committed:
Fix calculate reward function

A reward is given by picking up or putting down the predefined target objects. The task is still incomplete (target number of pickups is not handled)

beduffy commented on gym_ai2thor/envs/ai2thor_env.py:
```diff
@@ -78,6 +79,7 @@ def step(self, action, verbose=True):
         if not self.action_space.contains(action):
             raise error.InvalidAction(f'Action must be an integer between '
                                       f'0 and {self.action_space.n}!')
+        prev_state = deepcopy(self.event)
```
You're copying something that might be in megabytes in size just to compare if inventory has increased? Why not just deepcopy the inventory or even just keep track of the len of the list as I did? Why copy whole event?

fernandotorch committed:
Extend Readme for env and tasks

Added explanation on how to use the wrapper for defining environments and tasks using the config file

fernandotorch committed:
Fix tests

Bug in test_environment_runs due to refactoring
Changed the call for visualization since function is deprecated

beduffy commented on README.md:
```diff
@@ -1,35 +1,136 @@
- We also will work on representations which can transfer between tasks and environments.
+This project will focus primarily on the implementation and benchmark of different approaches to 
+domain and task transfer learning in reinforcement learning. The focus lies on a diverse set of 
+domestic robot tasks using [ai2thor](https://ai2thor.allenai.org/), a realistic household 3D 
```
domestic robot? A first person view and an integer action to just instantly pick up a cup is far from robotics. 

BUT. ai2thor gang did have a paper on transferring navigation to the real world. I don't mind too much keeping it but maybe it might be too much for now? Thoughts?

beduffy commented on README.md:
````diff
@@ -1,35 +1,136 @@
+                                                (step_n + 1) + (episode * max_episode_length),
+                                                 max_episode_length, time.time() - start))
+```
+
````
Whyyyyyyyy did you add the full example?

The whole point was to fit in in 6 lines for simplicity and understanding and grokking it in 0.3 seconds.
The rest is boilerplate.

beduffy replied:
Maybe you can convince me. Or maybe we can meet halfway by removing the last 4 lines

beduffy commented on README.md:
```diff
@@ -1,35 +1,136 @@
+ 
+For experimentation it is important to be able to make slight modifications of the environment 
+ without having to create a new config each time. The class `AI2ThorEnv` includes the keyword 
+ argument `config_dict`, that allows to input a python dictionary **in additon to** the config file 
```
in "addition" typo

beduffy commented on README.md:
````diff
@@ -1,35 +1,136 @@
+ ```
+ 
+For experimentation it is important to be able to make slight modifications of the environment 
+ without having to create a new config each time. The class `AI2ThorEnv` includes the keyword 
````
"new config file each time"

beduffy commented on README.md:
```diff
@@ -1,35 +1,136 @@
+For experimentation it is important to be able to make slight modifications of the environment 
+ without having to create a new config each time. The class `AI2ThorEnv` includes the keyword 
+ argument `config_dict`, that allows to input a python dictionary **in additon to** the config file 
+ that overrides the parameters described in the config.
```
To avoid repetition and better way of putting it, say:
(Careful: config_dict can override override parameters set in the config file"

fernandotorch replied:
Not careful, this is the only intended use. Not having to write everything but overriding one param. Do you agree?

beduffy replied:
it might not overwrite. 
it might write something new to the dict.
Something that might not be in the config files but can be added to all environments. This is obvious no?

beduffy commented on README.md:
```diff
@@ -1,35 +1,136 @@
+        super().__init__(kwargs)
+
+    def transition_reward(self, prev_state, post_state):
+        reward = -1 if prev_state == post_state else 1 
```
have you tested this? state is MASSIVE. lots of things change even if you walk into a wall. Bit silly. Bleh you could just put the pickup cup task here instead. It's very easy to move.

fernandotorch committed:
Fix typos and readme examples

Fixed example of how to create a task
Linked ai2thor tutorial as documentation source

beduffy committed:
Added config_files folder, read path correctly. removed __init__.py

Removed f-syntax

beduffy committed:
Changed from ini config files to json, implemented config_dict overwriting

Added 1 test case to check if warnings occur and overwriting works.
Moved start controller to end to do heavy stuff last.
Saved scene_id into var.

beduffy committed:
Added task_variation_random_walk.py fror showing config_dict functionality

removed max_episode_length

beduffy committed:
Added test case for cup picking, fixed objectType bug

Changed all step_n to step_num

fernandotorch committed:
Merge branch 'master' into ai2thor-wrapper

fernandotorch committed:
Fix minor errors

Check if the receptacle is closed before opening action.
Update readme and requirements
Reduce number of episodes in examples to 3 so that the code can be executed til the end in a reasonable time

fernandotorch committed:
Merge remote-tracking branch 'origin/ai2thor-wrapper' into ai2thor-wrapper

fernandotorch committed:
Add python version to readme

Ai2thor supports 2.7. and 3.5+ but we only work with 3.5+

fernandotorch committed:
Update tests

Renamed test_config for clarity
Actions changed to string

fernandotorch committed:
Refactor transition_reward from tasks

BaseTask method arguments simplified, refactored consistently in env as well.
Code from transition_reward simplified
